include/xsparse/levels/hashed.hpp

```diff
@@ -0,0 +1,167 @@
+                        const noexcept
+                    {
+                        return { wrapped_it->first, wrapped_it->second };
+                    }
+
+                    inline bool operator==(const iterator& other) const noexcept
+                    {
+                        return wrapped_it == other.wrapped_it;
+                    }
+
+                    inline iterator& operator++() noexcept
+                    {
+                        ++wrapped_it;
+                        return *this;
+                    }
+
+                    inline iterator& operator--() noexcept
+                    {
+                        --wrapped_it;
+                        return *this;
+                    }
+                };
+
+                explicit inline iteration_helper(typename BaseTraits::Level const& level,
+                                                 typename BaseTraits::I const i,
+                                                 typename BaseTraits::PKM1 const pkm1) noexcept
+                    : m_level(level)
+                    , m_pkm1(pkm1)
+                    , m_i(i)
+                {
+                }
+
+                inline iterator begin() const noexcept
+                {
+                    return iterator{ m_level.m_crd[m_pkm1].begin() };
+                }
+
+                inline iterator end() const noexcept
+                {
+                    return iterator{ m_level.m_crd[m_pkm1].end() };
+                }
+            };
+
+            iteration_helper iter_helper(typename BaseTraits::I i, typename BaseTraits::PKM1 pkm1)
+            {
+                return iteration_helper{ *static_cast<typename BaseTraits::Level*>(this), i, pkm1 };
+            }
+
+            hashed(IK size)
+                : m_crd_size(std::move(size))
+                , m_crd()
+            {
+            }
+
+            hashed(IK size, CrdContainer const& crd)
+                : m_crd_size(std::move(size))
+                , m_crd(crd)
+            {
+            }
+
+            hashed(IK size, CrdContainer&& crd)
+                : m_crd_size(std::move(size))
+                , m_crd(crd)
+            {
+            }
+
+            inline auto locate(typename BaseTraits::PKM1 pkm1, IK ik) const noexcept
+            {
+                auto it = m_crd[pkm1].find(ik);
+                return it != m_crd[pkm1].end() ? std::optional<PK>(it->second) : std::nullopt;
+            }
+
+            inline void insert_init(typename BaseTraits::IK szkm1) const noexcept
+            {
+                m_crd.resize(szkm1);
+            }
+
+            inline void insert_coord(typename BaseTraits::PKM1 pkm1, PK pk, IK ik) const noexcept
+            {
+                m_crd[pkm1][ik] = pk;
+            }
+
+            inline void insert_edges(typename BaseTraits::PKM1 pkm1) const noexcept
+            {
+                m_crd_size += m_crd[pkm1].size();
+            }
+
+        private:
+            PK m_crd_size;
```

Comment:
This is also needed in the constructors.

Reply:
Is the size of type IK also needed along with `m_crd_size`?

Reply:
No, that isn't needed.

Reply:
Done.

Reply:
Oh, my bad, I misunderstood your question, yes, size is needed.